src/vsr.zig

```diff
@@ -1098,3 +1102,221 @@ test "ViewChangeHeaders.view_for_op" {
+                }
+            } else {
+                break :command .outdated;
+            }
+        };
+
+        if (command_durable == .do_view_change and command_current == .do_view_change) {
+            assert(op_head_durable == op_head_current);
+            // Ensure that if we started a DVC before a crash, that we will resume sending the exact
+            // same DVC after recovery. (An alternative implementation would be to load the
+            // superblock's DVC headers (including gaps) into the journal during Replica.open(), but
+            // that is more complicated to implement correctly).
+            for (durable.headers.slice) |*header| headers.appendAssumeCapacity(header.*);
+            return;
+        }
+
+        // What is the relationship between two prepares?
+        const Chain = enum {
+            // The ops are sequential, and the hash-chain is valid.
+            chain_sequence,
+            // The ops are sequential, and the hash-chain is invalid.
+            chain_break,
+            // The ops are non-sequential, and belong to the same view.
+            // This gap never hides a break.
+            chain_view,
+            // The ops are non-sequential, and belong to the different views.
+            // Depending on the context, this gap may hide a break.
+            chain_gap,
+        };
+
+        // The DVC anchor: Within the log suffix following the anchor, we have additional
+        // guarantees about the state of the log headers which allow us to tolerate certain
+        // gaps (by locally guaranteeing that the gap does not hide a break).
+        const op_dvc_anchor = std.math.max(
+            options.op_checkpoint,
+            // +1: We may have a full pipeline, but not yet have performed any repair.
+            // In such a case, we want to send those pipeline_prepare_queue_max headers in
+            // the DVC, but not the preceding op (which may belong to a different chain).
+            // This satisfies the DVC invariant because the first op in the pipeline is
+            // "connected" to the canonical chain (via its "parent" checksum).
+            1 + op_head_current -| constants.pipeline_prepare_queue_max,
+        );
+
+        var suffix_done = false;
+        for (current.headers.constSlice()) |*header, i| {
+            const op = header.op;
+            const chain = chain: {
+                // Always include the head message.
+                if (i == 0) break :chain Chain.chain_sequence;
+
+                const child = headers.get(i - 1);
+                if (child.op == header.op + 1) {
+                    break :chain if (child.parent == header.checksum) Chain.chain_sequence else Chain.chain_break;
+                } else {
+                    break :chain if (child.view == header.view) Chain.chain_view else Chain.chain_gap;
+                }
+            };
+
+            if (current.log_view == current.view) {
```

Comment:
> Maybee `command_current == .start_view` to deduplicate the condition?

:+1: 

> Would also appreciate a comment on `const command_current:` explaining the semantics. I understand what `current.log_view == current.view` condition means (the current view is an active one), but I don't understand why that translates to the two commands.

Does this help?

```zig
        // The rules for generating DVCs and SVs differ. We use the current view numbers to
        // determine which is being generated:
        // - when `log_view < view`, generate a DVC
        // - when `log_view = view`, generate a SV
        const command_current: enum { start_view, do_view_change } =
            if (current.log_view == current.view) .start_view else .do_view_change;
        // Likewise, the durable view numbers identify whether the durable headers were from a past
        // DVC or SV. The durable headers are only useful if they are from the same view as our
        // current headers, though.
        const command_durable: enum { start_view, do_view_change, outdated } = command: {
            if (durable.log_view == current.log_view) {
                if (durable.log_view == durable.view) {
                    break :command .start_view;
                } else {
                    break :command .do_view_change;
                }
            } else {
                break :command .outdated;
            }
        };
```

> Maybe its better to pass command in, like we do in create_view_change_message, and assert?
> Or even, pass the command in for current, assert, and then use asserted logic to re-derive it for durable?

In `create_view_change_message` we assert:

```
            assert(command != .do_view_change or self.log_view < self.view);
            assert(command != .start_view or self.log_view == self.view);
```

Passing in the command there isn't strictly necessary — it can be derived entirely from the replica state. But passing in the command there enables a little extra verification of the call site.


I don't pass the command to `build()` because:
- We would have the `command` via the `create_view_change_message` call site, but the `view_durable_update` call site would have to re-derive it anyway, using the exact same method.
- (Less important:) I like that the `current`/`durable` parameters mirror each other. And `view`/`log_view` provides more information than the `command` alone.

Reply:
>Likewise, the durable view numbers identify whether the durable headers were from a past  DVC or SV.

This is still confusing to me, what polarity are we talking about? _sending_ DVC or SV (where we create VC message to send) or _receiving_ DVC or SV (where we replace our headers using received message[s], and then create VC message for `view_durable_update`)

Reply:
It is kind of both, but also they are sort of the same thing :sweat_smile: 

This function is used for all of the following:
- primary generating a SV to send to its backups
- primary generating a SV to write to its superblock
- primary/backup generating a DVC to send
- primary/backup generating a DVC to write to its superblock
- And this is a weird one: backups generating a SV which they write to their superblock. This is mostly for symmetry, since of course a backup will never send a SV message over the network. But since we only ever do this after receiving (and installing) a SV from the primary, it follows the same rules as the primary's SV generation.